midp/gfx.js

```diff
@@ -1600,22 +1582,25 @@ var currentlyFocusedTextEditor;
     };
 
     Native["com/nokia/mid/ui/TextEditor.setConstraints.(I)V"] = function(constraints) {
-        this.textEditor.setConstraints(constraints);
+        this.textEditor = TextEditorProvider.getEditor(constraints, this.textEditor, this.textEditorId);
     };
 
     Native["com/nokia/mid/ui/TextEditor.getConstraints.()I"] = function() {
-        return this.textEditor.getConstraints();
+        return this.textEditor.constraints;
     };
 
     Native["com/nokia/mid/ui/TextEditor.setFocus.(Z)V"] = function(shouldFocus) {
+        var promise;
         if (shouldFocus && (currentlyFocusedTextEditor != this.textEditor)) {
-            this.textEditor.focus();
+            promise = this.textEditor.focus();
             currentlyFocusedTextEditor = this.textEditor;
         } else if (!shouldFocus && (currentlyFocusedTextEditor == this.textEditor)) {
-            this.textEditor.blur();
+            promise = this.textEditor.blur();
             currentlyFocusedTextEditor = null;
+        } else {
+            promise = Promise.resolve();
```

Comment:
Natives that call asyncImpl don't have to do so in every case. The VM is smart enough to do the right thing if an "async native" that is listed in *yieldMap* sometimes returns a synchronous value. So this case should just return.